src/overlay/OverlayManagerImpl.cpp

```diff
@@ -138,17 +138,29 @@ OverlayManagerImpl::connectToImpl(PeerRecord pr)
 
 void
 OverlayManagerImpl::storePeerList(std::vector<std::string> const& list,
-                                  bool preferred)
+                                  bool setPreferred)
 {
     for (auto const& peerStr : list)
     {
         try
         {
             auto address = PeerBareAddress::resolve(peerStr, mApp);
-            auto pr = PeerRecord{address, mApp.getClock().now(), 0};
-            pr.setPreferred(preferred);
-            pr.resetBackOff(mApp.getClock());
-            pr.storePeerRecord(mApp.getDatabase());
+            auto pr = PeerRecord::loadPeerRecord(mApp.getDatabase(), address);
+            if (!pr)
+            {
+                pr = make_optional<PeerRecord>(address, mApp.getClock().now());
+            }
+            else
+            {
+                pr->resetBackOff(mApp.getClock());
+            }
+            if (setPreferred)
+            {
+                // we do not want to reset flags that were true in database
```

Comment:
Maybe I'm not clear enough here. What I wanted to say is:

* if that peer that we just loaded from database is indeed a preferred peer, but it was only discovered after authentication at some, then we do not set its preferred flag to false, because setPreferred is false
* if that peer was not stored in database with 'preferred' flag set to false, but before this run of stellar-core configuration was changed and it is now preferred, we want it to have that preferred flag set

So we are only preserving flags when setPreferred is false, when it is true we are force-setting those flags to true.